src/tuple17.lib/tuple17/Tuple.test.cpp

```diff
@@ -10,38 +10,202 @@
+    // static_assert(T::offset_of<Inner> == offset2);
 
-    ASSERT_EQ((g.of<double>()), 4.2);
-    ASSERT_EQ((g.of<int>()), 23);
-    ASSERT_EQ((g.of<float>()), 0);
+    static_assert(T::type_at<0> == type<Inner>);
+    static_assert(T::type_at<1> == type<char>);
+    static_assert(T::type_at<2> == type<Inner>);
+
+    static_assert(T::hasType<Inner>());
+    static_assert(T::hasType<char>());
+
+    static_assert(T::type_at<0> == type<Inner>);
+    static_assert(T::type_at<1> == type<char>);
+    static_assert(T::type_at<2> == type<Inner>);
+
+    auto t = T{};
+    static_assert(sizeof(t) == offsetT);
+
+    static_assert(std::tuple_size<T>::value == 3);
+    static_assert(type<std::tuple_element<0, T>::type> == type<Inner>);
+    static_assert(type<std::tuple_element<1, T>::type> == type<char>);
+    static_assert(type<std::tuple_element<2, T>::type> == type<Inner>);
 }
 
-TEST(Tuple, SingleArgCopy) {
-    using X = Tuple<int>;
+TEST(Tuple, access) {
+    using T = Tuple<char, int, double>;
+    auto t = T{};
+
+    t.at(index<1>) = 23;
+    t.of<double>() = 4.2;
+    EXPECT_EQ((t.of<int>()), 23);
+    EXPECT_EQ((t.at<2>()), 4.2);
+
+    EXPECT_EQ(std::get<char>(t), '\0');
+    EXPECT_EQ(std::get<int>(t), 23);
+    EXPECT_EQ(std::get<double>(t), 4.2);
+}
+
+TEST(Tuple, construction) {
+    using MultiArg = Tuple<char, int, double>;
+    auto m0 = MultiArg{};
+    auto m1 = m0; // copy
+    auto m2 = std::move(m1); // move copy
+
+    EXPECT_EQ(m0, m2);
+
+    using SingleArg = Tuple<int>;
+
+    SingleArg s0{23};
+    SingleArg s1(s0); // Test that it won't use the argument forward constructor
 
-    X a{23};
-    X b(a); // Test that it won't use the argument forward constructor
+    EXPECT_EQ(s0, s1);
+
+    using Ambiguous = Tuple<char, int, int>;
+    auto a0 = Ambiguous{};
+    // TODO CK: Tuple is sorted, should accept only sorted list
+    // Cannot assign due to ambiguous type index association int int
+    // auto a1 = Ambiguous{'c'};
+    // auto a2 = Ambiguous{'c', 23};
+    // auto a3 = Ambiguous{'c', 23, 32};
+}
+
+TEST(Tuple, broaden) {
+    using Small = Tuple<char, int, double>;
+    using Large = Tuple<int, double, char, float>;
+    auto s = Small{'c', 23, 4.2};
+    auto l = Large(s);
+
+    EXPECT_EQ((l.of<char>()), 'c');
+    EXPECT_EQ((l.of<double>()), 4.2);
+    EXPECT_EQ((l.of<int>()), 23);
+    EXPECT_EQ((l.of<float>()), 0);
+}
+
+TEST(Tuple, broadenTODO_2) {
+    using Small = Tuple<char, int>;
+    using Large = Tuple<int, float, char, int>;
+
+    auto s = Small{'c', 23};
+    // other order / bigger
+    // TODO CK: Why does this compile when  broaden chack1 does not compile?
+    auto l = Large(s);
+
+    EXPECT_EQ((l.of<char>()), 'c');
+    EXPECT_EQ((l.of<float>()), 0);
+
+    EXPECT_EQ((l.at<0>()), 23);
+    EXPECT_EQ((l.at<1>()), 0);
+    EXPECT_EQ((l.at<2>()), 'c');
+    // TODO CK: Value 23 - what is expected? I think it should prevent the creation as incompatible!
+    // ASSERT_EQ((l.at<3>()), 0);

```

Comment:
Yes this is weird but I don't want to spend time on thinking about these corner cases, until we have a use case.
Tuple should be fast and correct for a "custom" definition of input spaces. Strange behavior in unused corner cases, should not bother us.